build_tools/benchmarks/common/benchmark_definition.py

```diff
@@ -389,33 +389,127 @@ def from_json_object(json_object: Dict[str, Any]):
+
+
+def get_google_benchmark_latencies(
+    benchmark_json: Dict[str,
+                         Any]) -> Tuple[BenchmarkLatency, BenchmarkLatency]:
+  """Returns the Google Benchmark aggregate latencies.
+
+    Args:
+      benchmark_json: The JSON string or object returned by Google Benchmark.
+
+    Returns:
+      Real time and CPU time BenchmarkLatency.
+    """
+  real_time_object = dict(unit="ns")
+  cpu_time_object = dict(unit="ns")
+  metrics = ["mean", "median", "stddev"]
+  for case in benchmark_json["benchmarks"]:
+    if any(case["name"].endswith(f"real_time_{m}") for m in metrics):
+      if case["time_unit"] != "ns":
+        raise ValueError(f"Expected ns as time unit")
+      metric = case["name"].split("_")[-1]
+      real_time_object[metric] = int(round(case["real_time"]))
+      cpu_time_object[metric] = int(round(case["cpu_time"]))
+
+  # from_json_object implicitly validates that all metrics were found.
+  real_time = BenchmarkLatency.from_json_object(real_time_object)
+  cpu_time = BenchmarkLatency.from_json_object(cpu_time_object)
+  return real_time, cpu_time
+
+
+@dataclasses.dataclass
+class BenchmarkMetrics(object):
+  """An object describing the results from a single benchmark.
+
+  - real_time: the real time latency statistics returned by the benchmarking
+      framework.
+  - cpu_time: the cpu time latency statistics returned by the benchmarking
+      framework.
+  - raw_data: additional JSON-compatible raw results returned by the
+      benchmarking framework.
+  """
+  real_time: BenchmarkLatency
+  cpu_time: BenchmarkLatency
+  raw_data: Dict[str, Any]
+
+  def to_json_object(self) -> Dict[str, Any]:
+    return {
+        "real_time": self.real_time.to_json_object(),
+        "cpu_time": self.cpu_time.to_json_object(),
+        "raw_data": self.raw_data,
+    }
+
+  @staticmethod
+  def from_json_object(json_object: Dict[str, Any]):
+    return BenchmarkMetrics(
+        BenchmarkLatency.from_json_object(json_object["real_time"]),
+        BenchmarkLatency.from_json_object(json_object["cpu_time"]),
+        json_object["raw_data"],
+    )
+
+
+def parse_and_serialize_iree_benchmark_metrics(
+    results_filename: pathlib.Path, benchmark_stdout: str) -> BenchmarkMetrics:
+  """Extract benchmark metrics from the output of iree-benchmark-module.
+
+  Args:
+    results_filename: The path to serialize the BenchmarkMetrics to.
+    benchmark_stdout: The stdout of iree-benchmark-module with
+      --benchmark_format=json.
+
+  Returns:
+    A populated BenchmarkMetrics dataclass.
+  """
+  iree_benchmark_json = json.loads(benchmark_stdout)
+  real_time, cpu_time = get_google_benchmark_latencies(iree_benchmark_json)
+  benchmark_metrics = BenchmarkMetrics(
+      real_time=real_time,
+      cpu_time=cpu_time,
+      raw_data=iree_benchmark_json,
+  )
+  with open(results_filename, "w") as f:
+    f.write(json.dumps(benchmark_metrics.to_json_object()))
+  return benchmark_metrics
```

Comment:
It looks a little bit weird to me to 1. serialize and write out the metrics object but also 2. return the metrics object.

Maybe we just return the object and let the caller serialize and write out the file? This way we don't need to pass `results_filename` and this function won't have side effect.